File: app/src/main/java/com/github/opengles_android/common/Shader.java
package com.github.opengles_android.common;

import android.content.Context;
import android.opengl.GLES30;
import android.util.Log;

import java.util.HashMap;
import java.util.Map;

public class Shader {
    private final static String TAG = "Shader";
    private int mProgram;
    private Map<String, Object> mParamLocations = new HashMap<>();

    public Shader(Context context, String vertexFileName, String fragmentFileName) {
        this(Utils.getStringFromAssets(context, vertexFileName), Utils.getStringFromAssets(context, fragmentFileName));
    }

    public Shader(String vertexCode, String fragCode) {
        int vertexShader = loadShader(vertexCode, GLES30.GL_VERTEX_SHADER);
        int fragShader = loadShader(fragCode, GLES30.GL_FRAGMENT_SHADER);
        mProgram = createProgram(vertexShader, fragShader);
    }

    private int createProgram(int vShader, int fShader) {
        int program = GLES30.glCreateProgram();

        if (0 == program) {
            Log.e(TAG, "createProgram: Create program error!", new Throwable());
            return 0;
        }

        GLES30.glAttachShader(program, vShader);
        GLES30.glAttachShader(program, fShader);

        GLES30.glLinkProgram(program);

        int[] result = new int[1];
        GLES30.glGetProgramiv(program, GLES30.GL_LINK_STATUS, result, 0);

        if (0 == result[0]) {
            String log = GLES30.glGetProgramInfoLog(program);
            Log.e(TAG, "loadShader: Link program error!" + log, new Throwable());
            return 0;
        }

        return program;
    }

    private int loadShader(String code, int type) {
        int shader = GLES30.glCreateShader(type);
        GLES30.glShaderSource(shader, code);
        GLES30.glCompileShader(shader);

        int[] result = new int[1];
        GLES30.glGetShaderiv(shader, GLES30.GL_COMPILE_STATUS, result, 0);

        if (0 == result[0]) {
            String log = GLES30.glGetShaderInfoLog(shader);
            Log.e(TAG, "loadShader: Compile shader error!\n"
                    + "code: " + code + "\n"
                    + log, new Throwable());
            return 0;
        }

        return shader;
    }

    public void setInt(String name, int value) {
        GLES30.glUniform1i(getUniformLocation(name), value);
    }

    public void setMat4(String name, float[] value) {
        GLES30.glUniform4fv(getUniformLocation(name), 1, value, 0);
    }

    private int getUniformLocation(String name) {
        int location;

        if (null == mParamLocations.get(name)) {
            location = GLES30.glGetUniformLocation(mProgram, name);
            Log.i(TAG, "getUniformLocation, Set location, name: " + name + ", location: " + location);
            mParamLocations.put(name, location);
        } else {
            location = (int) mParamLocations.get(name);
        }

        return location;
    }

    public void use() {
        GLES30.glUseProgram(mProgram);
    }
}

File: app/src/main/cpp/Shader.cpp
//
// Created by xulinkai on 2021/10/24.
//

#include "Shader.h"

Shader::~Shader() {
    if (mProgram != 0) {
        glDeleteProgram(mProgram);
        mProgram = 0;
    }
}

Shader::Shader(const char *vertexCode, const char *fragCode) {
    InitShaderProgram(vertexCode, fragCode);
}

void Shader::SetMat4(const std::string& name, const glm::mat4& mat) {
    GLint location = GetUniformLocation(name);

    if (location != -1) {
        glUniformMatrix4fv(location, 1, GL_FALSE, glm::value_ptr(mat));
    }
}

void Shader::SetInt(const std::string& name, int value) {
    GLint location = GetUniformLocation(name);

    if (location > 0) {
        glUniform1i(location, value);
    }
}

GLint Shader::GetUniformLocation(const std::string& name) {
    if (mUniformLocations.find(name) != mUniformLocations.end()) {
        return mUniformLocations[name];
    }

    GLint location = glGetUniformLocation(mProgram, name.c_str());
    mUniformLocations[name] = location;

    return location;
}

GLuint Shader::LoadShader(const char* code, GLenum type) {
    GLuint shader = glCreateShader(type);

    if (0 == shader) {
        LOGE("glCreateShader %d failed!", type);
        return 0;
    }

    // Load the shader source
    glShaderSource(shader, 1, &code, nullptr);

    // Compile the shader
    glCompileShader(shader);

    // Check the compile status
    GLint status;
    glGetShaderiv(shader, GL_COMPILE_STATUS, &status);

    if (!status) {
        GLchar eLog[1024] = { 0 };
        glGetShaderInfoLog(shader, sizeof(eLog), NULL, eLog);
        LOGE("LoadShader, Error compiling the %d shader: '%s'", type, eLog);
        glDeleteShader(shader);
        return 0;
    }
    return shader;
}

void Shader::InitShaderProgram(const char *vertexCode, const char *fragCode) {
    mProgram = glCreateProgram();

    if (0 == mProgram) {
        LOGE("InitShaderProgram, create mShaderProgram failed!");
        return;
    }

    GLint vertexShader = LoadShader(vertexCode, GL_VERTEX_SHADER);
    GLint fragShader = LoadShader(fragCode, GL_FRAGMENT_SHADER);

    glAttachShader(mProgram, vertexShader);
    glAttachShader(mProgram, fragShader);

    glLinkProgram(mProgram);

    GLint linked;
    glGetProgramiv(mProgram, GL_LINK_STATUS, &linked);

    if (!linked) {
        GLchar eLog[1024] = { 0 };
        glGetProgramInfoLog(mProgram, sizeof(eLog), NULL, eLog);
        LOGE("InitShaderProgram, Error linking mShaderProgram: '%s'", eLog);
        glDeleteProgram(mProgram);
    }
}

void Shader::UseProgram() {
    glUseProgram(mProgram);
}
